### DISCUSSION TITLE
feature: How to extend the config of nvim-dap without copying the code?

### DISCUSSION THREAD

**Discussion by you-n-g** [LazyVim/LazyVim#700] [Q&A]
### Did you check the docs?

- [X] I have read all the LazyVim docs

### Is your feature request related to a problem? Please describe.

Hello
I'm configuring the nvim-dap.

I want to config Python for it.
So I'm planning to add config like this https://github.com/mfussenegger/nvim-dap/blob/master/doc/dap.txt#L218
into a function like the image below
![image](https://user-images.githubusercontent.com/465606/235611237-c7603b0b-51d9-4f01-a1fd-275843e990c2.png)

But the `config` attributed is already implemented in 
https://github.com/LazyVim/LazyVim/blob/main/lua/lazyvim/plugins/extras/dap/core.lua#L91

So how can I extend the configure (keep the current implementation and then extend my own config) without copying the code?

Thanks.

### Describe the solution you'd like

Support extending the configuration.

### Describe alternatives you've considered

NA

**Comment by dpetka2001** [comment]
There's already an external source for [python](https://github.com/mfussenegger/nvim-dap-python). You should use that.

**Comment by dpetka2001** [comment]
I have the following and it just works
```lua
return {
  -- Modify nvim-dap
  {
    "mfussenegger/nvim-dap",
    dependencies = {
      "mfussenegger/nvim-dap-python",
      config = function()
        require("dap-python").setup("~/.local/share/nvim/mason/packages/debugpy/venv/bin/python")
      end,
    },
  },

  -- Mason-nvim-dap
  {
    "jay-babu/mason-nvim-dap.nvim",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, {
        "python",
      })
    end,
  },
}
```
For further customization you should read the nvim-dap-python docs as I only use the default configuration.

**Comment by you-n-g** [comment]
@dpetka2001 thanks for your reply.
It works.
And I create a more general version based on your config. 
https://github.com/you-n-g/deploy/blob/a61b738c4d4f3b911f1d05266b86a1504175e558/configs/lazynvim/lua/plugins/nvim-dap.lua#L3
```lua
-- get the python path from the venv(like which python)
-- get the conda env path
local function get_python_path()
  -- TODO: this is only compatible with linux
  -- Implement a version that works with windows as well
  local path = vim.fn.system("which python")
  -- right trim the '\n' in   path
  path = string.gsub(path, "\n$", "")
  if path == "" then
    return nil
  end
  return path
end

-- local py_path =
return {
  {
    "mfussenegger/nvim-dap",
    dependencies = {
      "mfussenegger/nvim-dap-python",
      keys = {
        {
          "<leader>dG",
          function()
            local config = {
              configurations = {
                {
                  name = "vscode json launcher",
                  type = "python",
                  request = "launch",
                  cwd = vim.fn.getcwd(),
                  python = get_python_path(),
                  stopOnEntry = true,
                  -- console = "externalTerminal",
                  debugOptions = {},
                  program = vim.fn.expand("%:p"),
                -- TODO: select args in neovim
                  args = {},
                },
              },
            }
            config["configurations"][1]["justMyCode#json"] = "${justMyCode:true}"
            -- dump `config` to a json file
            local json = vim.fn.json_encode(config)
            -- create .vscode dir if not exists
            vim.fn.mkdir(".vscode", "p")
            local ok, file = pcall(io.open, ".vscode/launch.json", "w")
            if ok then
              file:write(json)
              print("Generated .vscode/launch.json")
            else
              print("Error writing to file: " .. file)
            end
          end,
          mode = "n",
          desc = "Generate .vscode/launch.json",
        },
      },
      config = function()
        require("dap-python").setup(get_python_path())
        require("dap.ext.vscode").load_launchjs()
      end,
    },
  },
}
```

**Comment by you-n-g** [comment]
For the topic, I have two different  naive implementation ideas that may support this feature.
- Make config accept a list of functions. Concatenate the list when merging the config.  Run them one-by-one when initialization.


- Support dummy dependencies. User can define some dependencies that only contains configuation without installing a new plugin.
For example
```lua
return {
  {
    "mfussenegger/nvim-dap",
    dependencies = {
      "nvim-dap-extra-conf",
      config = function()
        require("dap-python").setup(get_python_path())
        require("dap.ext.vscode").load_launchjs()
      end,
      dummy = true
    },
  },
}
```

**Comment by EmileVezinaCoulombe** [comment]
@you-n-g To extend your function `get_python_path` with **mason** and support for Windows users.
```lua
local function get_python_path()
  local mason_path = vim.fn.glob(vim.fn.stdpath("data") .. "/mason/")
  if mason_path ~= "" and (vim.fn.has("macunix") == 1 or vim.fn.has("unix") == 1) then
    return mason_path .. "packages/debugpy/venv/bin/python"
  elseif mason_path ~= "" and (vim.fn.has("win32") == 1) then
    return mason_path .. "packages\\debugpy\\venv\\Scripts\\python.exe"
  else
    return nil
  end
end
```

**Reply by dpetka2001** [reply]
Nice stuff, tnx for your implementation. One quick question. What is the difference between `glob()` and `expand()`? I've seen quite a few implementations using the latter as far as string paths are concerned.

**Reply by EmileVezinaCoulombe** [reply]
Good point, `glob` is OS agnostic so `vim.fn.glob(vim.fn.stdpath("data") .. "/mason/")` does work on Windows. `glob` try to get the path if it exists but `expand` will only concatenate the strings. I have updated the function to reflect this.
